fix(ivy): unable to project into multiple slots with default selector

devversion opened the pull request:
With View engine it was possible to declare multiple default projection
definitions and to programmatically project nodes into the slots.

e.g.
```html
<ng-content></ng-content>
<span>Spacer</span>
<ng-content></ng-content>
```

Using `ViewContainerRef#createComponent` allowed projecting
nodes into one of the both projection defs (through index)

This no longer works with Ivy as the projectionDef instruction only
retrieves a list of selectors instead of also retrieving entries for
reserved projection slots which appear when using the default
selector multiple times (as seen above).

Two possible solutions to fix this: https://hackmd.io/s/Sy2kQlgTE

This PR implements solution (2) where we completely follow the
semantics of content projection in view engine.

pkozlowski-opensource commented on packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts:
```diff
@@ -1209,6 +1211,54 @@ describe('compiler compliance', () => {
             result.source, ComplexComponentDefinition, 'Incorrect ComplexComponent definition');
       });
 
+      it('should support multi-slot content projection', () => {
```
Could we update the name to make it clear that we are dealing with multiple "wildcard" slots here?

devversion replied:
Good point. will do that!

pkozlowski-opensource commented on packages/core/test/render3/content_spec.ts:
```diff
@@ -823,7 +823,7 @@ describe('content projection', () => {
     const Child = createComponent('child', function(rf: RenderFlags, ctx: any) {
       if (rf & RenderFlags.Create) {
-        ΔprojectionDef();
+        ΔprojectionDef(['*']);
```
Maybe `[*]` should be the default and could be omitted to save on generated code and memory allocation?

devversion replied:
That is a good point. I was slightly afraid of doing this initially as it makes the `projectionDef` instruction less explicit but given this is such a common case it seems to make sense (especially if we could avoid allocating a new array every time)

pkozlowski-opensource commented on tools/public_api_guard/core/core.d.ts:
```diff
@@ -1313,7 +1313,7 @@ export declare type ΔPipeDefWithMeta<T, Name extends string> = PipeDef<T>;
 export declare function Δprojection(nodeIndex: number, selectorIndex?: number, attrs?: TAttributes): void;
 
-export declare function ΔprojectionDef(selectors?: CssSelectorList[]): void;
+export declare function ΔprojectionDef(projectionSlots: ProjectionSlots): void;
```
Yeh, I would probably mark `projectionSlots` as optional

pkozlowski-opensource commented:
@devversion did a first pass on it, LGTM generally speaking, thnx! 
Just left few minor comments / nits.

pkozlowski-opensource reviewed (approved):
LGTM! 

CI is red but it looks like a flake....

devversion commented:
@pkozlowski-opensource Thanks for reviewing. Restarted the build. everything passes now

JoostK commented on packages/compiler-cli/test/compliance/r3_compiler_compliance_spec.ts:
```diff
@@ -1209,6 +1209,54 @@ describe('compiler compliance', () => {
+            vars: 0,
+            template: function Cmp_Template(rf, ctx) {
+              if (rf & 1) {
+                i0.ΔprojectionDef($c0$);
```
I am surprised CI is green, as this seems to have been rebased on top of the frog eyes, whereas this test still asserts for `Δ`.

devversion replied:
Good catch. This is actually a bug within the `expectEmit` utility function. I'll provide a fix in a separate PR.

devversion replied:
Done.

devversion committed:
fix(ivy): unable to project into multiple slots with default selector

With View engine it was possible to declare multiple projection
definitions and to programmatically project nodes into the slots.

e.g.

```html
<ng-content></ng-content>
<ng-content></ng-content>
```

Using `ViewContainerRef#createComponent` allowed projecting
nodes into one of the projection defs (through index)

This no longer works with Ivy as the `projectionDef` instruction only
retrieves a list of selectors instead of also retrieving entries for
reserved projection slots which appear when using the default
selector multiple times (as seen above).

In order to fix this issue, the Ivy compiler now passes all
projection slots to the `projectionDef` instruction. Meaning that
there can be multiple projection slots with the same wildcard
selector. This allows multi-slot projection as seen in the
example above, and it also allows us to match the multi-slot node
projection order from View Engine (to avoid breaking changes).

It basically ensures that Ivy fully matches the View Engine behavior
except of a very small edge case that has already been discussed
in FW-886 (with the conclusion of working as intended).

Read more here: https://hackmd.io/s/Sy2kQlgTE

mhevery commented:
http://test/OCL:250816978:BASE:250816986:1559273777771:4edbf03a

mhevery commented:
Merge to master, but does not apply cleanly to patch. Please create a new PR which targets patch.
